### PR TITLE
Update Namespace default SubnetSize at runtime

### PR DESCRIPTION
**Pull request by wenqiq**
Update Namespace default SubnetSize at runtime

1. There is a `VPCNetworkConfiguration` in the Cluster with `defaultSubnetSize: 32`, now we change the defaultSubnetSize from 32 to 16:

```
kubectl get VPCNetworkConfiguration wenqi2  -oyaml
apiVersion: crd.nsx.vmware.com/v1alpha1
kind: VPCNetworkConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"crd.nsx.vmware.com/v1alpha1","kind":"VPCNetworkConfiguration","metadata":{"annotations":{},"name":"wenqi2"},"spec":{"defaultSubnetSize":32,"nsxProject":"/orgs/default/projects/nsx_operator_e2e_test","privateIPs":["172.191.0.0/16"],"vpcConnectivityProfile":"/orgs/default/projects/nsx_operator_e2e_test/vpc-connectivity-profiles/default"}}
  creationTimestamp: "2024-08-21T06:43:51Z"
  generation: 5
  name: wenqi2
  resourceVersion: "8588470"
  uid: 46a04316-5d48-4961-9bd1-5d77d0973c2d
spec:
  defaultSubnetSize: 16
  nsxProject: /orgs/default/projects/nsx_operator_e2e_test
  privateIPs:
  - 172.191.0.0/16
  vpcConnectivityProfile: /orgs/default/projects/nsx_operator_e2e_test/vpc-connectivity-profiles/default
status:
  vpcs:
  - lbSubnetPath: /orgs/default/projects/nsx_operator_e2e_test/vpcs/wenqi2-6e46aef6-f247-4742-aabd-1203f4dc493b/subnets/_AVI_SUBNET--LB
    name: wenqi2-6e46aef6-f247-4742-aabd-1203f4dc493b
```

2. Create a namespace `wenqi-test-0826-7` with annotation: `nsx.vmware.com/vpc_network_config: wenqi2`
```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    ako.vmware.com/tenant-name: nsx_operator_e2e_test
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{"nsx.vmware.com/vpc_network_config":"wenqi2"},"name":"wenqi-test-0826-7"}}
    nsx.vmware.com/vpc_network_config: wenqi2
  creationTimestamp: "2024-08-26T08:00:01Z"
  labels:
    kubernetes.io/metadata.name: wenqi-test-0826-7
  name: wenqi-test-0826-7
  resourceVersion: "8584753"
  uid: bd17637e-954e-46c4-806f-b50c455e2893
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

3. The new `subnetset` in the namespace `wenqi-test-0826-7` with the `IPV4SUBNETSIZE` equal to the `defaultSubnetSize` in `VPCNetworkConfiguration`.
```
kubectl get subnetset -n wenqi-test-0826-7
NAME          ACCESSMODE   IPV4SUBNETSIZE   IPADDRESSES
pod-default   PrivateTGW   16               
vm-default    Private      16   
```

4. Update the `defaultSubnetSize` in `VPCNetworkConfiguration`

image

5. Create a new `Subnetset` in Namespace 
image

6. Check the `ipv4SubnetSize` value in new `Subnetset`
image

### CODE REVIEW

File: pkg/controllers/subnet/subnet_controller.go [hunk 1]
```diff
@@ -64,23 +64,20 @@ func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+
+			vpcNetworkConfig := r.VPCService.GetVPCNetworkConfigByNamespace(obj.Namespace)
+			if vpcNetworkConfig == nil {
+				err := fmt.Errorf("operate failed: cannot get configuration for Subnet CR")
+				log.Error(nil, "failed to find VPCNetworkConfig for Subnet CR", "subnet", req.NamespacedName, "namespace %s", obj.Namespace)
+				updateFail(r, ctx, obj, "")
+				return ResultRequeue, err
+			}
+
+			obj.Spec.IPv4SubnetSize = vpcNetworkConfig.DefaultSubnetSize
```
Comment (dantingl) [inline pkg/controllers/subnet/subnet_controller.go:80]: I think you also need to check `obj.Spec.IPv4SubnetSize == 0` as user can manually create a Subnet/SubnetSet CR with size, right?

Comment (wenqiq) [reply pkg/controllers/subnet/subnet_controller.go]: Updated.

File: pkg/controllers/subnet/subnet_controller.go [hunk 2]
```diff
@@ -64,22 +64,20 @@ func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
-					obj.Spec.IPv4SubnetSize = vpcNetworkConfig.DefaultSubnetSize
-				}
+			vpcNetworkConfig := r.VPCService.GetVPCNetworkConfigByNamespace(obj.Namespace)
+			if vpcNetworkConfig == nil {
+				err := fmt.Errorf("operate failed: cannot get configuration for Subnet CR")
+				log.Error(nil, "failed to find VPCNetworkConfig for Subnet CR", "subnet", req.NamespacedName, "namespace %s", obj.Namespace)
+				updateFail(r, ctx, obj, "")
+				return ResultRequeue, err
+			}
+			if obj.Spec.IPv4SubnetSize == 0 {
```
Comment (dantingl) [inline pkg/controllers/subnet/subnet_controller.go:79]: put line 72-78 in 79 condition?

File: pkg/controllers/subnetset/subnetset_controller.go [hunk 3]
```diff
@@ -61,20 +61,19 @@ func (r *SubnetSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+				log.Error(err, "operate failed, would retry exponentially", "subnet", req.NamespacedName)
+				updateFail(r, ctx, obj, "")
+				return ResultRequeue, err
+			}
+
+			if obj.Spec.AccessMode == "" {
+				obj.Spec.AccessMode = v1alpha1.AccessMode(v1alpha1.AccessModePrivate)
+			}
+			if obj.Spec.IPv4SubnetSize == 0 {
+				obj.Spec.IPv4SubnetSize = vpcNetworkConfig.DefaultSubnetSize
```
Comment (dantingl) [inline pkg/controllers/subnetset/subnetset_controller.go:76]: same here

**Review by dantingl — COMMENTED** [review]
You need another test case:
after NS creation, update subnet size in VPCNetworkConfiguration CR 
create new Subnet and SubnetSet CR without setting subnet size
check whether the size in Subnet/SubnetSet CR is updated

**Review by zhengxiexie — APPROVED** [review]
LGTM

**Review by dantingl — APPROVED** [review]

**Pull request merged**